Boost: Refactor the pre-WordPress part of the Page Cache

donnchawp opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->
This PR started out by removing the 0 byte index.html files left by garbage collection, but is now a refactor of the Page Cache storage system.
It uses "path actions" to handle the different actions we perform on files and directories, such as deleting and rebuilding, and filtering out expired files.

Tests were added to cover most of the changes.

Fixes HOG-23
And probably will fix HOG-22 too as the number of cached files and directories is reduced.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Fix the $iterator in the cache cleanup so it returns directories as well as files.
* Remove the REBUILD check when doing GC so directories can be cleaned up.
* 

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Apply PR
* Enable Page Cache and make sure some pages are cached.
* Clear cache from the settings page and confirm that the blog's directory in boost-cache/cache/ is gone, empty, or almost empty. (random requests from the internet may populate pages)
* Set the cache TTL to 600 seconds by adding `define( 'JETPACK_BOOST_CACHE_DURATION', 600 );` to your wp-config.php
* Wait for an hour for garbage collection or run the "jetpack_boost_cache_garbage_collection" through the wp-crontrol plugin. Confirm that any cache file older than an the TTL is gone. Old index.html files may still exist if there are newer cache files in that directory or sub directories.
* Make sure garbage collection works and removes old files. Set the TTL to 10 seconds. Run the scheduled job manually.
* Use the jetpack_boost_clear_page_cache_url action to try to get out of boost-cache/cache/ by using URLs with ../../ in them.
* Unpublish a post and confirm it doesn't delete the entire cache.
* Tail the error log and make sure no file operation warnings appear.

donnchawp committed:
We need to remove directories when rebuilding too

GC can delete rebuild files, leaving empty directories. Removing this
check makes sure those directories are deleted.

donnchawp committed:
Return directoroes in iterator too.

donnchawp committed:
changelog

donnchawp committed:
Add tests for Filesyste_Utils.php

donnchawp committed:
Fix the type of the $type parameter as static analysis checks failed

donnchawp committed:
Checks seem to be stalled. Kick them?

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/fix-deleting-directories

donnchawp committed:
Updating PHAN baseline

haqadn reviewed (dismissed):
Nothing to change. Just testing.

haqadn reviewed (commented):
The change makes sense but I faced an issue which I am not sure how it's happening.

When I click clear cache button, it still leaves the index.html files. However, Deactivating and then reactivating the cache module clears the cache nicely.

I am confused because the code path is almost identical for both.

donnchawp committed:
Delete index.html files when deleting everything

If that file is left over, then the directory can't be deleted

donnchawp commented:
It was [this line](https://github.com/Automattic/jetpack/pull/42789/files#diff-8622fc9f8fd99d72c80dd909f33ff702982db2cba210df1ae403052d0d39f35eL47) that excluded index.html. When I was testing it on my server I had removed that check. I could have sworn I removed it locally too before checking in my code but I didn't.
Clearing the cache will now delete the entire directory for your hostname.

donnchawp commented:
I was able to clear the cache on a site with over 200MB of files in about 11 seconds. Another site with 64MB of files was cleared in about 2 seconds.

LiamSarsfield reviewed (dismissed):
LGTM, works as expected!
I noticed the log files for boost-cache still exist after uninstalling the plugin `boost-cache/logs`, though this is most likely out of scope for this for this PR.

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/class-filesystem-utils.php:
```diff
@@ -198,10 +201,6 @@ public static function gc_expired_files( $directory, $file_ttl, $action = self::
-			return $count;
-		}
-
 		// If the directory is empty after processing its files, delete it.
```
Seems like we are handling index.html in a special way and only deleting it when deleting the directory. Check the definition of `is_dir_empty`:

https://github.com/Automattic/jetpack/blob/dc83c8e07bc72912269718224c70ace6f8bfd2e9/projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/class-filesystem-utils.php#L314C2-L321C3

For some reason that approach is not working. I think we should fix that instead.

haqadn committed:
Refactor walk_directory and garbage collection

haqadn committed:
Decouple file-storage from cache

haqadn committed:
Reuse delete and rebuild path actions

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/class-boost-cache.php:
```diff
@@ -329,12 +313,12 @@ public function rebuild_on_comment_post( $comment_id, $comment_approved, $commen
 			if ( isset( $parameters['cookies'] ) && ! empty( $parameters['cookies'] ) ) {
 				$filename = trailingslashit( get_permalink( $post->ID ) ) . Filesystem_Utils::get_request_filename( $parameters );
-				$this->storage->invalidate( $filename, Filesystem_Utils::DELETE_FILE );
+				$this->rebuild_page( $filename );
```
This is broken atm. It should delete the specific file with specific parameters. This need to be addressed before the PR is merged.

donnchawp replied:
PR has been updated. This `rebuild_page` should have been `delete_page` too. It's because the cached page was generated by a visitor with cookies who just left a comment. The cookies will be unique to that visitor.

donnchawp committed:
Fix this typo and it's one file

donnchawp committed:
Delete the cache file for this user as they left a comment

We need to delete this file instead of rebuild as the cache file is only
used by that particular user because of their cookies.

donnchawp committed:
Add $parameters parameter to rebuild and delete page functions

This allows those functions to delete a single page visited by a
particular visitor with cookies.

haqadn commented:
Tests are failing at the moment.

donnchawp committed:
Apply the action to the root $path too.

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/boost-cache-actions.php:
```diff
@@ -44,15 +45,15 @@
 function jetpack_boost_delete_cache_for_home() {
 	$boost_cache = new Boost_Cache();
-	$boost_cache->invalidate_cache_for_front_page( Filesystem_Utils::DELETE_ALL );
+	$boost_cache->delete_page( home_url() );
```
I was trying to make sense of the action and function name and decided it should only delete the homepage. But, the previous implementation was deleting the whole cache.

donnchawp replied:
This needs to replicate what the original `invalidate_cache_for_front_page` did. If the homepage shows a static page it needs to delete the "posts" page too. The stable version of the plugin deletes the "homepage" post too, but it might be enough to delete home_url() for that. I'll fix this.

donnchawp replied:
I just updated the PR to delete the posts page too.

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/boost-cache-actions.php:
```diff
@@ -62,7 +63,7 @@ function jetpack_boost_delete_cache_for_home() {
 function jetpack_boost_delete_cache_for_url( $url ) {
 	$boost_cache = new Boost_Cache();
-	$boost_cache->invalidate_cache_for_url( $url, Filesystem_Utils::DELETE_ALL );
+	$boost_cache->delete_page( $url );
```
Does this make sense to be `delete_recursive`? I am unsure, asking based on old implementation.

donnchawp replied:
Yes, it should be recursive. Before it was modified recently, in version 3.12.1 it did a DELETE_ALL by default. Also, it would allow you to delete or rebuild specific cache files, but I doubt anyone used that functionality.

This function is used by the `jetpack_boost_clear_page_cache_url` hook to allow developers to delete pages. However, [the example given](https://github.com/Automattic/jetpack/blob/update/boost/fix-deleting-directories/projects/plugins/boost/docs/Cache_API.md#delete-the-cache-for-a-specific-url) uses https://example.com/ so a recursive delete would clear the entire cache. I need to update those docs to show a post or page, and not the homepage.

donnchawp replied:
I updated the docs and changed this to `delete_recursive`.

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/boost-cache-actions.php:
```diff
@@ -71,6 +72,32 @@ function jetpack_boost_delete_cache_for_url( $url ) {
+
 	$boost_cache = new Boost_Cache();
-	$boost_cache->invalidate_cache_by_post_id( (int) $post_id, Filesystem_Utils::DELETE_ALL );
+	$boost_cache->delete_page( $permalink );
```
I guess this should be `delete_recursive` as well.

donnchawp replied:
Yeah. A post may have paged comments, so yeah. This is now using delete_post_cache which does a recursive delete.

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/class-filesystem-utils.php:
```diff
@@ -45,6 +45,8 @@ public static function iterate_directory( $path, Path_Action $action ) {
 			$count += $action->apply_to_path( $file );
 		}
 
+		$count += $action->apply_to_path( new SplFileInfo( $path ) );
```
I tried this and found it may produce errors in GC due to the directory not existing. I think we can either leave the directory alone or check for directory exists in the delete action.

donnchawp replied:
I tested this by deleting the base hostname directory and triggering a GC and I didn't see any errors, so the extra checks we've added since, have fixed this.

haqadn committed:
Re-use some code

haqadn committed:
Remove unused method

haqadn committed:
Refactor the storage interface

haqadn committed:
Refactor rebuilding

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/fix-deleting-directories

donnchawp committed:
Merge branch 'update/boost/fix-deleting-directories' of github.com:Automattic/jetpack into update/boost/fix-deleting-directories

donnchawp committed:
Fix this comment approved check modified in f1609ea3d6b4e4a8c4a0dd43a04cedd1ac53ca6c

donnchawp committed:
Fix for PHAN complaints

donnchawp committed:
Fix the $created_before passed to Filter_Order

It was a duration, but needs to be a timestamp

donnchawp committed:
apply_to_path takes a SplFileInfo, not a string

donnchawp committed:
Add delete_empty_dir and call that from delete and rebuild actions

donnchawp committed:
Shortcircuit the file age check if it's a directory

donnchawp committed:
Just delete it, the Older Filter is much older than rebuild TTL.

donnchawp committed:
The Older Filter is much longer than the rebuild TTL.

donnchawp committed:
Fix counts to fix tests

donnchawp committed:
Various PHAN checks fixed.

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/fix-deleting-directories

donnchawp committed:
Update the example here to a specific page.

Clearing the homepage would clear the entire cache.

donnchawp committed:
Fix double slashes on normalized paths.

donnchawp committed:
Make sure $path has a trailingslash instead of adding an extra one

donnchawp committed:
Check if file, before deleting it

donnchawp committed:
Recursively delete urls

donnchawp committed:
If the homepage is static, delete the posts page

haqadn committed:
Update tests

haqadn committed:
Add tests for path actions

haqadn committed:
Add tests for path actions

donnchawp committed:
Suppress warnings from these commands

In a caching system where garbage collection can delete cache files, and
other processes can too, it can happen that a file is deleted while a
chunk of code is still running in a different process that expects it to
be there.

donnchawp committed:
Return true to make testing easier

haqadn committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/fix-deleting-directories

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/class-cache-preload.php:
```diff
@@ -205,17 +204,20 @@ public function handle_post_update( int $post_id ) {
-	public function handle_cache_invalidation( string $path, string $type ) {
-		if ( $type === Filesystem_Utils::DELETE_ALL ) {
+	public function handle_cache_invalidation( string $path, string $type, string $scope ) {
+		if ( $path === home_url() && $scope === 'recursive' ) {
```
I'm curious why do we check if the path is the home url?

Do we make the assumption that if it's the home page that we're invalidating for (and we're doing it recursively), the whole cache was invalidated?

I think I get it but I'm not a 100% sure. Would a comment make sense here?

dilirity replied:
Do we need to worry if slashes are present or not in `$path` or is the path always sanitized?

donnchawp replied:
The `handle_cache_invalidation` function is used by the `jetpack_boost_invalidate_cache_success` action which is executed by the `invalidate_cache_success` function.
The function is used by the various delete/rebuild functions in Boost_Cache, which are called with home_url() or permalinks, although jetpack_boost_delete_cache_for_url() does call it with $url.
That is fired by an action, `jetpack_boost_clear_page_cache_url`, so $url could be anything. For `handle_cache_invalidation`, that doesn't matter because it's only used for comparison, but I need to check where that's sanitized before deleting.

donnchawp replied:
> I'm curious why do we check if the path is the home url?

Yeah, there is a comment a few lines down saying "If the cache is invalidated for all files, schedule preload for all Cornerstone Pages."

> Do we need to worry if slashes are present or not in $path or is the path always sanitized?
I was going to remove any "//" but that might introduce security issues. If someone passes through a url with "//" separating items in the URL, then it won't match and in this case, if they clear "http://example.com//" then it won't schedule a cornerstone refresh.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/boost-cache-actions.php:
```diff
@@ -44,15 +44,23 @@
+	if ( get_option( 'show_on_front' ) === 'page' ) {
+		$posts_page_id = get_option( 'page_for_posts' ); // posts page
+		if ( $posts_page_id ) {
+			$boost_cache->delete_recursive( get_permalink( $posts_page_id ) );
```
Having an id for the option, doesn't mean the page actually exists.

I know it's an edge case, but if the blog page gets deleted there's no permalink.

donnchawp replied:
I just tested that. If the page is deleted, then show_on_front won't be "page" any more, so this is ok.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/class-boost-cache.php:
```diff
@@ -231,40 +231,25 @@ public function ob_callback( $buffer ) {
-				Logger::debug( 'invalidate_cache_for_front_page: deleting front page cache' );
-				$this->invalidate_cache_for_post( get_post( $front_page_id ), $action );
+				Logger::debug( 'rebuild_front_page: deleting front page cache' );
+				$this->rebuild_post_cache( get_post( $front_page_id ) );
```
Can this check if what we're sending is an actual post? `get_post` might return `null` if the post has been deleted.

donnchawp replied:
I think it's safe enough to do this as WordPress takes care of the setting if the page is deleted.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/class-boost-cache.php:
```diff
@@ -231,40 +231,25 @@ public function ob_callback( $buffer ) {
-				Logger::debug( 'invalidate_cache_for_front_page: deleting posts page cache' );
-				$this->invalidate_cache_for_post( get_post( $posts_page_id ), $action );
+				Logger::debug( 'rebuild_front_page: deleting posts page cache' );
+				$this->rebuild_post_cache( get_post( $posts_page_id ) );
```
Same as the comment above about the post being deleted.

donnchawp replied:
WordPress takes care of it, so this seems to be safe.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/class-filesystem-utils.php:
```diff
@@ -13,14 +16,67 @@ class Filesystem_Utils {
 	const REBUILD_FILE_EXTENSION = '.rebuild.html'; // The extension used for rebuilt files.
+
+	/**
+	 * Iterate over a directory and apply an action to each file.
```
Nitpick: I think using file when it refers to a file path is a bit confusing. The code refers paths everywhere, but then it passes files instead of file paths.

donnchawp replied:
It just means the files in the directory, although I see what you mean. The actions take `SplFileInfo` objects where everywhere else it's a string.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/class-filesystem-utils.php:
```diff
@@ -13,14 +16,67 @@ class Filesystem_Utils {
+	 * @param Path_Action $action - The action to apply to each file.
+	 * @return int|Boost_Cache_Error - The number of files processed, or Boost_Cache_Error on failure.
+	 */
+	public static function iterate_directory( $path, Path_Action $action ) {
```
Nitpick: The name `iterate_directory` doesn't sit right with me for some reason. It's a given that it iterates, but the most important part is applying the action.

donnchawp replied:
`process_directory_recursively` perhaps? Or `apply_directory_action` ?

dilirity replied:
I think we can ignore this. Let's not make this take any longer than necessary 😅

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/class-filesystem-utils.php:
```diff
@@ -13,14 +16,67 @@ class Filesystem_Utils {
+	 * @return int|Boost_Cache_Error - The number of files processed, or Boost_Cache_Error on failure.
 	 */
-	public static function walk_directory( $path, $type ) {
+	public static function iterate_files( $path, Path_Action $action ) {
```
The initial description of this method overlaps with `iterate_directory`.

I have the same nitpick for the name.

donnchawp replied:
Both functions do something similar, but maybe `process_single_directory` ? Or `process_files` ?

dilirity replied:
Let's ignore this for now.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/path-actions/class-filter-older.php:
```diff
@@ -0,0 +1,35 @@
+<?php
```
Nitpick: I really liked the comments on everything else in page cache 😅

donnchawp replied:
Ha, I'll add them to these actions then. :)

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/path-actions/class-filter-older.php:
```diff
@@ -0,0 +1,35 @@
+
+class Filter_Older implements Path_Action {
+	private $timestamp;
+	private $sub_action;
```
Curious: Why is this `$sub_action` instead of `$action`?

donnchawp replied:
Adnan wrote this, but maybe because this filter is an action itself, so $sub_action clearly separates the action to perform from the filter.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/path-actions/class-simple-delete.php:
```diff
@@ -0,0 +1,46 @@
+	 * Delete a file or directory.
+	 *
+	 * @param SplFileInfo $file The file or directory to delete.
+	 * @return int The number of files or directories deleted.
```
This mentions returning `int`, but `delete_file` returns a boolean.

donnchawp replied:
Good catch, I'll make apply_to_path return 1 after deleting.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/path-actions/class-simple-delete.php:
```diff
@@ -0,0 +1,46 @@
+	 * @param SplFileInfo $file The file or directory to delete.
+	 * @return int The number of files or directories deleted.
+	 */
+	public function apply_to_path( SplFileInfo $file ) {
```
I think this can also return `null` if the directory isn't empty.

donnchawp replied:
I'll make it return 0. Other bits of code count the number of files affected.

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/storage/class-file-storage.php:
```diff
@@ -152,33 +163,39 @@ private function sanitize_path( $path ) {
 
-		if ( $response === null ) {
-			return new Boost_Cache_Error( 'no-cache-files-to-delete', 'No cache files to delete.' );
+		$action = new Simple_Delete();
```
Nitpick: I know it's a little thing, but if we put this in the `else` we wouldn't always have to create an instance and then sometimes override it.

donnchawp replied:
I was thinking of doing that, but forgot to change that. Thanks for the reminder. :)

dilirity reviewed (changes requested):
Great work on this! I couldn't find any issues while testing.

I have a few notes/nitpicks and we can merge after that.

donnchawp committed:
Handle clearing cache when a post is unpublished.

The permalink is /?p=X instead, but get_sample_permalink() can return
the correct URL.

donnchawp committed:
Sanitize the path and check it's contained.

Protect the other items in wp-content/boost-cache/

donnchawp committed:
Add comments to path actions

donnchawp committed:
Return integers when deleting or nothing happens

donnchawp committed:
It's one or the other.

donnchawp committed:
Updated PHAN baseline

donnchawp committed:
Use get_page_link() to find a page permalink.

donnchawp committed:
Updates to deleting/rebuilding homepage

* Never recursively delete/rebuild the homepage when deleting a post
* Rebuild the home_url() even when it's a static page

donnchawp committed:
Use the Boost_Cache_Utils one here since it's pre-WordPress

haqadn commented on projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/class-boost-cache.php:
```diff
@@ -466,15 +473,27 @@ public function rebuild_all() {
 	public function delete_post_cache( $post ) {
 		$post_path = $this->get_post_path_for_invalidation( $post );
-		if ( $post_path ) {
+		if ( null === $post_path ) {
+			return;
+		}
+
+		if ( Boost_Cache_Utils::trailingslashit( $post_path ) !== Boost_Cache_Utils::trailingslashit( home_url() ) ) {
 			$this->delete_recursive( $post_path );
+		} else {
+			$this->delete_page( $post_path );
```
👍🏼
